vochain/account.go

```diff
@@ -175,108 +214,502 @@ func (v *State) GetAccount(address common.Address, isQuery bool) (*Account, erro
 	return &acc, acc.Unmarshal(raw)
 }
 
-// VerifyAccountBalance extracts an account address from a signed message, and verifies if
-// there is enough balance to cover an amount expense
-func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (bool, common.Address, error) {
-	address, err := ethereum.AddrFromSignature(message, signature)
+// AccountFromSignature extracts an address from a signed message and returns an account if exists
+func (v *State) AccountFromSignature(message, signature []byte) (common.Address, *Account, error) {
+	pubKey, err := ethereum.PubKeyFromSignature(message, signature)
+	if err != nil {
+		return common.Address{}, nil, fmt.Errorf("cannot extract public key from signature: %w", err)
+	}
+	address, err := ethereum.AddrFromPublicKey(pubKey)
 	if err != nil {
-		return false, address, err
+		return common.Address{}, nil, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	if address == types.EthereumZeroAddressBytes {
+		return types.EthereumZeroAddressBytes, nil, fmt.Errorf("invalid address")
 	}
 	acc, err := v.GetAccount(address, false)
 	if err != nil {
-		return false, address, fmt.Errorf("VerifyAccountWithAmmount: %v", err)
+		return common.Address{}, nil, fmt.Errorf("cannot get account: %v", err)
 	}
 	if acc == nil {
-		return false, address, nil
+		return common.Address{}, nil, ErrAccountNotFound
 	}
-	return acc.Balance >= amount, address, nil
+	return address, acc, nil
 }
 
-func (v *State) SetAccountInfoURI(address common.Address, infoURI string) error {
-	if infoURI == "" {
-		return fmt.Errorf("cannot set an empty URI")
+func (v *State) SetAccountInfoURI(accountAddress, txSender common.Address, infoURI string) error {
+	acc, err := v.GetAccount(accountAddress, false)
+	if err != nil {
+		return err
 	}
-	var acc Account
-	v.Tx.Lock()
-	defer v.Tx.Unlock()
-	raw, err := v.Tx.DeepGet(address.Bytes(), AccountsCfg)
-	if err != nil && !errors.Is(err, arbo.ErrKeyNotFound) {
+	if acc == nil {
+		return ErrAccountNotFound
+	}
+	SetAccountInfoCost, err := v.TxCost(models.TxType_SET_ACCOUNT_INFO, false)
+	if err != nil {
 		return err
-	} else if err == nil {
-		if err := acc.Unmarshal(raw); err != nil {
+	}
+	if accountAddress == txSender {
+		acc.InfoURI = infoURI
+		acc.Balance -= SetAccountInfoCost
+		acc.Nonce++
+		accBytes, err := acc.Marshal()
+		if err != nil {
 			return err
 		}
+		v.Tx.Lock()
+		defer v.Tx.Unlock()
+		return v.Tx.DeepSet(accountAddress.Bytes(), accBytes, AccountsCfg)
+	}
+	sender, err := v.GetAccount(txSender, false)
+	if err != nil {
+		return err
+	}
+	if sender == nil {
+		return ErrAccountNotFound
+	}
+	sender.Balance -= SetAccountInfoCost
```

Comment:
I think all these balance/account operations should be perform with TransferBalance() and send it to 0xFFFF

Also I'm not sure the balance should be handled at this level... it feels to me that the transfers should stay in some other TX related layer